packages/flutter_driver/lib/src/driver/driver.dart

```diff
@@ -908,44 +872,30 @@ void restoreVmServiceConnectFunction() {
 
 /// Waits for a real Dart VM service to become available, then connects using
 /// the [VMServiceClient].
-///
-/// Times out after 30 seconds.
 Future<VMServiceClientConnection> _waitAndConnect(String url) async {
-  final Stopwatch timer = Stopwatch()..start();
-
-  Future<VMServiceClientConnection> attemptConnection() async {
-    Uri uri = Uri.parse(url);
-    if (uri.scheme == 'http')
-      uri = uri.replace(scheme: 'ws', path: '/ws');
-
+  Uri uri = Uri.parse(url);
+  if (uri.scheme == 'http')
+    uri = uri.replace(scheme: 'ws', path: '/ws');
+  int attempts = 0;
+  while (true) {
     WebSocket ws1;
     WebSocket ws2;
     try {
-      ws1 = await WebSocket.connect(uri.toString()).timeout(_shortTimeout(connectionTimeoutMultiplier));
-      ws2 = await WebSocket.connect(uri.toString()).timeout(_shortTimeout(connectionTimeoutMultiplier));
+      ws1 = await WebSocket.connect(uri.toString());
```

Comment:
Surely `WebSocket.connect` will fail in case of error?

I'm not sure what the purpose of the timeouts here were, originally. TCP itself has timeouts which should presumably kick in if something is going to time out. I would expect the regular failure mode here to be that the peer refused the connection (i.e. port not yet open).

Reply:
If the port's outright closed yes, but I've also seen connection attempts that just hang.

Reply:
They shouldn't hang forever, right? The kernel has a timeout on TCP connection attempts (derived from the TCP spec) which we should now be honoring.